bin/amphtml-update.py

```diff
@@ -772,6 +775,22 @@ def UnicodeEscape(string):
 	"""
 	return ('' + string).encode('unicode-escape')
 
+def AddMandatoryOf( spec_dict, attribute_spec, constraint ):
+	"""Gets the spec dictionary, with mandatory_*of constraints possibly added.
+
+	Args:
+		spec_dict: A dictionary of the attr_spec values to possibly add the constraint to.
+		attribute_spec: The attribute spec in which to look for the mandatory_*of constraint.
+		constraint: A string of the mandatory_*of constraint, like 'mandatory_anyof'.
+	Returns:
+		The spec_dict, with the constraint possibly added.
+	"""
+	if attribute_spec.HasField(constraint):
+		mandatory_of = getattr(attribute_spec, constraint).lstrip('[').rstrip(']').split(',')
```

Comment:
Why are the square brackets being stripped? I believe `[` should rather be replaced with `data-amp-bind-` if I understand this right.

Reply:
Yeah, this stripping of square brackets isn't very elegant. 


Those calls are there because the `.protoascii` files have [this format](https://github.com/ampproject/amphtml/blob/b765e57b8748e31de55fc8249ed0540500107997/extensions/amp-addthis/validator-amp-addthis.protoascii#L34) for `mandatory_oneof` and `mandatory_anyof`:

```
mandatory_oneof: "['data-product-code', 'data-widget-id']"
```

Without the `lstrip()` and `rstrip()` calls above, here's what it looks like:

```diff
diff --git a/includes/sanitizers/class-amp-allowed-tags-generated.php b/includes/sanitizers/class-amp-allowed-tags-generated.php
index cd8679638..db20c21b2 100644
--- a/includes/sanitizers/class-amp-allowed-tags-generated.php
+++ b/includes/sanitizers/class-amp-allowed-tags-generated.php
@@ -948,8 +948,8 @@ class AMP_Allowed_Tags_Generated {
                                'attr_spec_list' => array(
                                        'data-product-code' => array(
                                                'mandatory_oneof' => array(
-                                                       'data-product-code',
-                                                       'data-widget-id',
+                                                       '[\'data-product-code',
+                                                       'data-widget-id\']',
                                                ),
                                        ),

```

Still, there could be a better way to handle that.

Reply:
I see. Strange that the value is a string and not already parsed.

Reply:
Yeah, it's a little weird. Maybe a regex would be more elegant, but it might be a little more code.